contracts/evm/contracts/adapters/CentralizedConnection.sol

```diff
@@ -0,0 +1,148 @@
+    mapping(string => mapping(uint256 => bool)) receipts;
+    address private xCall;
+    address private adminAddress;
+
+    event Message(string targetNetwork, int256 sn, bytes _msg);
+
+    modifier onlyAdmin() {
+        require(msg.sender == this.admin(), "OnlyRelayer");
+        _;
+    }
+
+    function initialize(address _relayer, address _xCall) public initializer {
+        xCall = _xCall;
+        adminAddress = _relayer;
+    }
+
+    /**
+     @notice Sets the fee to the target network
+     @param networkId String Network Id of target chain
+     @param messageFee Integer ( The fee needed to send a Message )
+     @param responseFee Integer (The fee of the response )
+     */
+    function setFee(
+        string calldata networkId,
+        uint256 messageFee,
+        uint256 responseFee
+    ) external onlyAdmin {
+        messageFees[networkId] = messageFee;
+        responseFees[networkId] = responseFee;
+    }
+
+    /**
+     @notice Gets the fee to the target network
+    @param to String Network Id of target chain
+    @param response Boolean ( Whether the responding fee is included )
+    @return fee Integer (The fee of sending a message to a given destination network )
+    */
+    function getFee(
+        string memory to,
+        bool response
+    ) external view override returns (uint256 fee) {
+        uint256 messageFee = messageFees[to];
+        if (response == true) {
+            uint256 responseFee = responseFees[to];
+            return messageFee + responseFee;
+        }
+        return messageFee;
+    }
+
+    /**
+     @notice Sends the message to a specific network.
+     @param sn : positive for two-way message, zero for one-way message, negative for response
+     @param to  String ( Network Id of destination network )
+     @param svc String ( name of the service )
+     @param sn  Integer ( serial number of the message )
+     @param _msg Bytes ( serialized bytes of Service Message )
+     */
+    function sendMessage(
+        string calldata to,
+        string calldata svc,
+        int256 sn,
+        bytes calldata _msg
+    ) external payable override {
+        require(msg.sender == xCall, "Only Xcall can call sendMessage");
+        uint256 fee;
+        if (sn >= 0) {
+            fee = this.getFee(to, true);
+        } else {
+            fee = this.getFee(to, false);
+        }
+        require(msg.value >= fee, "Fee is not Sufficient");
+        emit Message(to, sn, _msg);
+    }
+
+    /**
+     @notice Sends the message to a xCall.
+     @param srcNetwork  String ( Network Id )
+     @param sn Integer ( serial number of the message )
+     @param _msg Bytes ( serialized bytes of Service Message )
+     */
+    function recvMessage(
+        string memory srcNetwork,
+        uint256 sn,
+        bytes calldata _msg
+    ) public onlyAdmin {
+        require(!receipts[srcNetwork][sn], "Duplicate Message");
+        receipts[srcNetwork][sn] = true;
```

Comment:
I believe, since the sn is always 0 for one-way xcall messages, using this criterion for identifying duplicate messages is not feasible.
@AntonAndell